validator/client/src/main/java/tech/pegasys/teku/validator/client/ActiveKeyManager.java

```diff
@@ -214,57 +217,107 @@ DeleteKeyResult deleteExternalValidator(final Validator activeValidator) {
+  }
+
+  private Set<BLSPublicKey> filterExternallyImportedPubKeys(
+      List<Pair<BLSPublicKey, PostKeyResult>> importResults) {
+    return importResults.stream()
+        .filter(
+            pubKeyPostKeyResultPair ->
+                pubKeyPostKeyResultPair.getValue().getImportStatus().equals(ImportStatus.IMPORTED))
+        .map(Pair::getKey)
+        .collect(Collectors.toSet());
   }
 
-  private PostKeyResult importValidatorFromKeystore(
+  private Pair<Optional<BLSPublicKey>, PostKeyResult> importValidatorFromKeystore(
       final String keystoreString,
       final String password,
       final Optional<SlashingProtectionImporter> slashingProtectionImporter) {
     final KeyStoreData keyStoreData = KeyStoreLoader.loadFromString(keystoreString);
     try {
       keyStoreData.validate();
     } catch (KeyStoreValidationException ex) {
-      return PostKeyResult.error(ex.getMessage());
+      return Pair.of(Optional.empty(), PostKeyResult.error(ex.getMessage()));
     }
 
     return validatorLoader.loadLocalMutableValidator(
         keyStoreData, password, slashingProtectionImporter);
   }
 
   @Override
-  public List<PostKeyResult> importExternalValidators(final List<ExternalValidator> validators) {
-    final List<PostKeyResult> importResults = new ArrayList<>();
+  public List<PostKeyResult> importExternalValidators(
+      final List<ExternalValidator> validators,
+      final Optional<DoppelgangerDetector> maybeDoppelgangerDetector) {
+    final List<Pair<BLSPublicKey, PostKeyResult>> importResults = new ArrayList<>();
     boolean reloadRequired = false;
     for (ExternalValidator v : validators) {
       try {
         importResults.add(
             validatorLoader.loadExternalMutableValidator(v.getPublicKey(), v.getUrl()));
-        if (importResults.get(importResults.size() - 1).getImportStatus()
+        if (importResults.get(importResults.size() - 1).getValue().getImportStatus()
             == ImportStatus.IMPORTED) {
           reloadRequired = true;
         }
       } catch (Exception e) {
-        importResults.add(PostKeyResult.error(e.getMessage()));
+        importResults.add(Pair.of(v.getPublicKey(), PostKeyResult.error(e.getMessage())));
       }
     }
     if (reloadRequired) {
-      validatorTimingChannel.onValidatorsAdded();
+      maybeDoppelgangerDetector
+          .map(
+              doppelgangerDetector ->
+                  doppelgangerDetector
+                      .performDoppelgangerDetection(filterExternallyImportedPubKeys(importResults))
+                      .thenAccept(
+                          doppelgangerDetected -> {
+                            if (!doppelgangerDetected.isEmpty()) {
+                              System.exit(1);
+                            }
+                          }))
+          .orElse(SafeFuture.COMPLETE)
+          .thenAccept(__ -> validatorTimingChannel.onValidatorsAdded())
+          .ifExceptionGetsHereRaiseABug();
```

Comment:
i think for external signers we need to be super careful. some people (against our recommendation) do run active/active, and this would cause them problems if they turn dopelganger on..

Ultimately, that's kind of why an external signer is in use in the first place, so downtime can be minimized moving keys etc, because the signing records are safe...

Potentially we don't run this protection on dopelganger, or we potentially do something clever, but i think ditto for the system.exit probably not being what we want...

Reply:
Again I think if you've chosen to enable doppelgänger, you're explicitly asking us to keep you safe even if that means missing some attestations when keys are first added or on restart. If you're running active/active or want to minimise downtime while moving validators you shouldn't enable doppelgänger.

Reply:
Again, the only other option I see here instead of shutting down the VC would be to remove the detected doppelganger keys.
I think we should make it clear that the doppelganger detection is "not perfect" and that running it clearly means missing some duties (which means penalties). That should be an accepted side effect.